### PR TITLE
Add index to `history_date` field to improve performance.

### LINKED ISSUE #565: Add index to `history_date` field
Methods `get_next_record()` and `get_prev_record()` filter table rows by `history_date` field. Adding an index to this field can improve performance of these methods

```
extra_fields = {
    ...
    "history_date": models.DateTimeField(db_index=True)
    ...
}
```

### LINKED ISSUE #758: Queryset as_of history performance issues
**Describe the bug**
The current implementation of a queryset as_of does the checks for changes (And removal of "-" history types) purely in a python set & generator - looping over every single item in the original queryset.

On sufficiently large querysets, this can take a significant amount of time to get a full set of historical models for a queryset at a given point in time.

**To Reproduce**
Have a large (Hundreds of thousands of records or more) queryset, and attempt to work on the entire queryset (For example, generating a set of diffs from a point of time to current over a whole set)

On my data, with ~47,000 items and ~80,000 historical rows, getting a set of changes for one filtered set can take up to 20 minutes.

(I'm aware of the other ticket to add filtering to these querysets and am planning on making a pull request for that in the future)

**Expected behavior**
Generation of the history should be done utilising the database where possible for significant speed gains.

**Environment (please complete the following information):**
 - OS: Mac / Ubuntu 18.04 Docker
 - Django Simple History Version: 2.8.0 ~ latest
 - Django Version: 3.0.3
 - Database Version: PostgreSQL 11

### PR DESCRIPTION
**Pull request by jeking3**
## Description

The field `history_date` has never been indexed, yet it is heavily relied on for lookups.  I followed the discussion in #565 and #729.  I split the latest() fix out of #729 and placed it into #862 so that the change to indexing history_date could be done in isolation.  Given the critical importance of indexing on `history_date` I have made it configurable but as an opt-out, not an opt-in; hopefully that is an acceptable compromise.

The as_of queries use a combination of the history_date and model pk, so a composite index on those two fields should also yield gains.  The `SIMPLE_HISTORY_DATE_INDEX` can be set to `composite` to enable this behavior.

## Related Issue

This closes #565.
This closes #729.
This closes #758.

## Motivation and Context

There are _many_ queries that filter on history_date, many using `lte` or `ge` type of queries.  Lack of an index can lead to a table scan, or at least needing to load in every record that meets the other filters.  The database should be doing this work.  With this change the work in #729 is completed and closes #565.  I have a project where I constrain the history_date to be between two timepoints and this eliminates loading all the Historical records just to find them.

## How Has This Been Tested?

The unit tests exercise `history_date` sufficiently, however I added a couple more for the new branches.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have run the `make format` command to format my code
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] I have added my name and/or github handle to `AUTHORS.rst`
- [X] I have added my change to `CHANGES.rst`
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by jeking3** [comment]
One job failed due to not having the fix in #862.  Once that is merged and this is rebased, the issue will resolve.

File: simple_history/models.py
Comment (tim-schilling) [inline simple_history/models.py:488]: I believe this change would necessitate a DB change via a migration. If this is enabled, does Django complain about model changes that don't have migrations created for it?

**Review by tim-schilling — COMMENTED** [review]
Does this approach allow for a developer to change their mind about wanting to have an index? If they use the library and have the index created, will the library determine how to drop that index?

I'd vote to have the index on the field exist by default then leave it up to the developer to add their own migration within their project to remove it if they'd like, avoiding the settings. I think this approach allows the library to sidestep the above scenario, reducing the scope of work. The documentation could even include the migration for dropping the index.

**Comment by jeking3** [comment]
Django manages indexes.  If you change the settings and run "make migrations", indexes on your historical tables will change appropriately.  There is no need for custom migrations.  If you switch from "enabled" to "composite" your django-generated migration will A) drop the single column index; then B) make a composite index.

File: simple_history/models.py
Comment (jeking3) [reply simple_history/models.py:488]: See below, it does not.

**Comment by jeking3** [comment]
@tim-schilling I added a response to your questions, please review.

**Comment by tim-schilling** [comment]
@jeking3 thank you for answering my questions. It looks like I was wrong in my thinking, I appreciate the clarifications. I'll take another look at the PR when I have time later this week. In the meantime, it does look like you committed the merge conflict syntax in `changes.rst`.

**Comment by jeking3** [comment]
I resolved the CHANGES.rst conflict this morning after merging someone else's PR.

**Comment by jeking3** [comment]
Aha, I see, yes there were a couple commits in my branch; I rebased before squashing them, and that crept in.  Fixed that up.

**Comment by jeking3** [comment]
Just rebased on master to fix up the codecov report.

File: CHANGES.rst
Comment (tim-schilling) [inline CHANGES.rst:15]: Is this correct?

**Review by tim-schilling — CHANGES_REQUESTED** [review]
I'd suggest making the changes to `changes.rst` outside of the scope of adding indexes on `history_date` in a new PR.

The code otherwise looks good! I will appreciate the work here a lot.

File: CHANGES.rst
Comment (jeking3) [reply CHANGES.rst:15]: Yes it is correct.  I merged a couple PRs that did not have entries in CHANGES.rst so I added them here.  I will clean this up so you can approve it.

**Comment by jeking3** [comment]
@tim-schilling I minimized the changes to CHANGES.rst per your request.

**Review by tim-schilling — APPROVED** [review]
Thanks for that last round of changes. I feel more comfortable in reviewing this now.

**Comment by imhmdb** [comment]
@jeking3 when can we expect this to be released? :(
it's been too long since it was merged

thanks for your amazing work btw <3

**Pull request merged**
